extensions-builtin/Lora/network_oft.py

```diff
@@ -55,26 +54,24 @@ def __init__(self,  net: network.Network, weights: network.NetworkWeights):
         elif is_other_linear:
             self.out_dim = self.sd_module.embed_dim
 
-        if self.is_kohya:
-            self.constraint = self.alpha * self.out_dim
-            self.num_blocks = self.dim
-            self.block_size = self.out_dim // self.dim
-        elif self.is_boft:
+        self.num_blocks = self.dim
+        self.block_size = self.out_dim // self.dim
+        self.constraint = (1 if self.alpha is None else self.alpha) * self.out_dim
+        if self.is_R:
             self.constraint = None
-            self.boft_m = weights.w["oft_diag"].shape[0]
-            self.block_num = weights.w["oft_diag"].shape[1]
-            self.block_size = weights.w["oft_diag"].shape[2]
+            self.block_size = self.dim
+            self.num_blocks = self.out_dim // self.dim
+        elif self.is_boft:
+            self.boft_m = self.oft_blocks.shape[0]
+            self.num_blocks = self.oft_blocks.shape[1]
+            self.block_size = self.oft_blocks.shape[2]
             self.boft_b = self.block_size
-            #self.block_size, self.block_num = butterfly_factor(self.out_dim, self.dim)
-        else:
-            self.constraint = None
-            self.block_size, self.num_blocks = factorization(self.out_dim, self.dim)
 
     def calc_updown(self, orig_weight):
         oft_blocks = self.oft_blocks.to(orig_weight.device)
         eye = torch.eye(self.block_size, device=oft_blocks.device)
 
-        if self.is_kohya:
+        if not self.is_R:
             block_Q = oft_blocks - oft_blocks.transpose(1, 2) # ensure skew-symmetric orthogonal matrix
```

Comment:
@KohakuBlueleaf  I used 2.1.0.post3 to train a new boft and diag-oft 
boft got an error during inference 
![56M0QF 3B{W$IZR~9I%YMG3](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/26623230/7e2fcbf1-31c1-4423-8c4f-c448c51d5a21)

Reply:
> @KohakuBlueleaf I used 2.1.0.post3 to train a new boft and diag-oft boft got an error during inference ![56M0QF 3B{W$IZR~9I%YMG3](https://private-user-images.githubusercontent.com/26623230/306660741-7e2fcbf1-31c1-4423-8c4f-c448c51d5a21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MjcyNzcsIm5iZiI6MTcwODUyNjk3NywicGF0aCI6Ii8yNjYyMzIzMC8zMDY2NjA3NDEtN2UyZmNiZjEtMzFjMS00NDIzLThjNGYtYzQ0OGM1MWQ1YTIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE0NDkzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5NmI1MjRjNWU5YzY3NTI4ZWVlM2E0OTU3ODM4MjM3Y2U4ZWE2MWYyNWZjYTExMTYxMzY2NjE1ZGU1OWI3ZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Fdm-0r_uoZ5WD8sMSz8rpwMsUNx2TcwSO4TOks5hbOo)

Got it

Reply:
> @KohakuBlueleaf I used 2.1.0.post3 to train a new boft and diag-oft boft got an error during inference ![56M0QF 3B{W$IZR~9I%YMG3](https://private-user-images.githubusercontent.com/26623230/306660741-7e2fcbf1-31c1-4423-8c4f-c448c51d5a21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MjczNTksIm5iZiI6MTcwODUyNzA1OSwicGF0aCI6Ii8yNjYyMzIzMC8zMDY2NjA3NDEtN2UyZmNiZjEtMzFjMS00NDIzLThjNGYtYzQ0OGM1MWQ1YTIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE0NTA1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViZWVkOWY2M2JlYWJjNTg3ZDkwZDI5OGRlOWIyZDFiMWEyNTQxYTg0YzZkMDViMzczNWU5MmViMzBhMDAwNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pejKmXdG7lIW8Mi-b4ANUsND20uYb2k3D5yC_vjpa0o)

Try new Commit

Reply:
> Thanks , boft looks like work, And I found that diag-oft needs to specify "constrain" in training, otherwise it doesn’t take effect. Is it a bug? ![xyz_grid-0000-528576219](https://private-user-images.githubusercontent.com/26623230/306698582-4e6981ef-6d38-4c17-97e0-56a10aeffbc3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MzQwMDIsIm5iZiI6MTcwODUzMzcwMiwicGF0aCI6Ii8yNjYyMzIzMC8zMDY2OTg1ODItNGU2OTgxZWYtNmQzOC00YzE3LTk3ZTAtNTZhMTBhZWZmYmMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE2NDE0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmMzM5ZWFlOWFmNjkzMDY2Y2FjNThkMWUzZTUxOWIwMDE1YjYwMjQyODJiZjA0YzliZmUyZDE5MjAwMmNlY2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LiSpNznHsJXjgSxLmZZHGTTuVhp21bLCbt05lx3ccMc)

It is bug

Reply:
@Exist-c fixed

Reply:
Noticed the same thing so went to find if anything has been updated, but after pulling the changes there is an error about 'MultiHeadAttention' object has no attribute 'weight'

send help pls ![image](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/15861396/c572e407-59b4-405b-8d5d-7af46a4551af)